src/IO/DataImporter/DataFileReaderActions.hpp

```diff
@@ -0,0 +1,197 @@
+ * must the the same that was encoded into the `observers::ArrayComponentId`
+ * used to register the elements.
+ */
+template <typename ImporterOptionsGroup, typename FieldTagsList,
+          typename CallbackAction, typename CallbackComponent>
+struct ReadElementData {
+  using const_global_cache_tags =
+      tmpl::list<Tags::DataFileName<ImporterOptionsGroup>,
+                 Tags::VolumeDataSubgroup<ImporterOptionsGroup>,
+                 Tags::ObservationValue<ImporterOptionsGroup>>;
+
+  template <typename ParallelComponent, typename DataBox,
+            typename Metavariables, typename ArrayIndex,
+            Requires<db::tag_is_retrievable_v<Tags::RegisteredElements,
+                                              DataBox>> = nullptr>
+  static void apply(DataBox& box,
+                    Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/,
+                    const gsl::not_null<CmiNodeLock*> node_lock) noexcept {
+    Parallel::lock(node_lock);
+    {
+      // The scoping is to close the file before unlocking
+      h5::H5File<h5::AccessType::ReadOnly> h5file(
+          Parallel::get<Tags::DataFileName<ImporterOptionsGroup>>(cache));
+      constexpr size_t version_number = 0;
+      const auto& volume_file = h5file.get<h5::VolumeData>(
+          "/" + Parallel::get<Tags::VolumeDataSubgroup<ImporterOptionsGroup>>(
+                    cache),
+          version_number);
+      const auto observation_id = volume_file.find_observation_id(
+          Parallel::get<Tags::ObservationValue<ImporterOptionsGroup>>(cache));
+      // Read the tensor data for all elements at once, since that's how it's
+      // stored in the file
+      tuples::tagged_tuple_from_typelist<FieldTagsList> all_tensor_data{};
+      tmpl::for_each<FieldTagsList>([
+        &all_tensor_data, &volume_file, &observation_id
+      ](auto field_tag_v) noexcept {
+        using field_tag = tmpl::type_from<decltype(field_tag_v)>;
+        auto& tensor_data = get<field_tag>(all_tensor_data);
+        for (size_t i = 0; i < tensor_data.size(); i++) {
+          tensor_data[i] = volume_file.get_tensor_component(
+              observation_id,
+              db::tag_name<field_tag>() + tensor_data.component_suffix(
+                                              tensor_data.get_tensor_index(i)));
+        }
+      });
+      // Retrieve the information needed to reconstruct which element the data
+      // belongs to
+      const auto all_grid_names = volume_file.get_grid_names(observation_id);
+      const auto all_extents = volume_file.get_extents(observation_id);
+      // Distribute the tensor data to the registered elements
+      for (auto& element_and_name : get<Tags::RegisteredElements>(box)) {
+        const CkArrayIndex& raw_element_index =
+            element_and_name.first.array_index();
+        // Check if the parallel component of the registered element matches the
```

Comment:
I don't understand; why would multiple components register their elements?

Reply:
Every element from any component that will receive data at some point must register (same for observers). Then the importer can be invoked any number of times with any file name or observation ID, and will distribute the data to the invoking component’s elements.

In other words, since the importer doesn’t know ahead of time which component will invoke it, the elements must uniquely identify themselves with their ID and their component.

Reply:
ah okay; please augment your test to show a case where multiple components read data

Reply:
done, please take a look at the fixup